### PR TITLE
Refactor cluster and machine actuators

### LINKED ISSUE #10: Implement MachineStatus for the machine actuator
Once the API is stable, consider moving from storing the status in annotations to the Machine `ProviderStatus` field

### LINKED ISSUE #19: Refactor loading credentials to actuator
As per #17, we could move this to the controller package and pass this to the actuators

### PR DESCRIPTION
**Pull request by justaugustus**
**What this PR does / why we need it**:

This re-approaches some of the changes in #42 in smaller chunks.
- Add Azure clients to actuator package
- Implement unified scope in actuator
- Wire services to use actuator scope clients 
- Wire cluster and machine actuators to use scope clients
- Update clusterctl, manager, and dependencies
- Update cluster generation tools, manifests, and instructions

Closes #10, closes #19 

_This also moves the project to using `kind`, instead of minikube. Huzzah! (thanks @BenTheElder)_

**Special notes for your reviewer**:

~Currently a work-in-progress, but it could use a first-pass review.~
/assign @tariq1890 

**Release note**:

```release-note
Refactor cluster and machine actuators
```

### CODE REVIEW

**Comment by justaugustus** [comment]
The next step I'll be doing here is modifying the existing functions within the actuator package to use the Azure clients instantiated in `scope.go`, instead of the referencing the old ones in `pkg/cloud/azure/services/{compute,network,resources}/`

**Comment by justaugustus** [comment]
(As a note, there are several types in `pkg/apis/azureprovider/v1alpha1/types.go` that reference AWS-specific properties. I've marked them as `TODO`s and will clean that up with a follow-up PR.)

File: pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// AzureMachineProviderStatus is the type that will be embedded in a Machine.Status.ProviderStatus field.
+// It containsk Azure-specific status information.
```
Comment (tariq1890) [inline pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go]:
```suggestion
// It contains Azure-specific status information.
```
Was this manually added or done by the codegen?

File: pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+// +k8s:openapi-gen=true
+type AzureMachineProviderStatus struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// VMID is the instance ID of the machine created in Azure
```
Comment (tariq1890) [inline pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go]: NIT: Let's have a period at the end of a sentence.

Comment (tariq1890) [reply pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go]: I know we haven't followed this in other PRs. We can start the norm from here on :)

File: pkg/apis/azureprovider/v1alpha1/types.go [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// AzureResourceReference is a reference to a specific Azure resource by ID
+type AzureResourceReference struct {
+	// ID of resource
+	// +optional
+	ID *string `json:"id,omitempty"`
```
[the remark this answers is not in the document]
Comment (tariq1890) [reply pkg/apis/azureprovider/v1alpha1/types.go]: Second this!

File: pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go]: 2019 since this is a new file
```suggestion
Copyright 2019 The Kubernetes Authors.
```

File: pkg/cloud/azure/actuators/scope.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline pkg/cloud/azure/actuators/scope.go]:
```suggestion
Copyright 2019 The Kubernetes Authors.
```

File: pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go (hunk 2 above)
Comment (justaugustus) [reply pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go]: Fixed!

File: pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go (hunk 1 above)
Comment (justaugustus) [reply pkg/apis/azureprovider/v1alpha1/azuremachineproviderstatus_types.go]: Manually. typo on my part. All fixed.

File: pkg/apis/azureprovider/v1alpha1/types.go [hunk 6]
```diff
@@ -0,0 +1,391 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline pkg/apis/azureprovider/v1alpha1/types.go]:
```suggestion
Copyright 2019 The Kubernetes Authors.
```

New file. New year!!

File: pkg/cloud/azure/actuators/clients.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline pkg/cloud/azure/actuators/clients.go]:
```suggestion
Copyright 2019 The Kubernetes Authors.
```

File: pkg/apis/azureprovider/v1alpha1/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (justaugustus) [reply pkg/apis/azureprovider/v1alpha1/types.go]: Thanks for the heads-up on this! I'm going to opt to assess this, along with the rest of the types, in a follow-up PR as a lot of this code is not currently exercised.

File: cmd/versioninfo/versioninfo.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline cmd/versioninfo/versioninfo.go]:
```suggestion
Copyright 2019 The Kubernetes Authors.
```

File: cmd/versioninfo/BUILD [hunk 9]
```diff
@@ -0,0 +1,17 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
```
Comment (tariq1890) [inline cmd/versioninfo/BUILD]: We haven't added Bazel to this project. If we are looking to do that, I think we should do that in another PR

File: cmd/versioninfo/versioninfo_test.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline cmd/versioninfo/versioninfo_test.go]:
```suggestion
Copyright 2019 The Kubernetes Authors.
```

File: config/crds/azureprovider_v1alpha1_azureclusterproviderstatus.yaml [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    plural: azureclusterproviderstatuses
+  scope: Namespaced
+  validation:
+    openAPIV3Schema:
+      properties:
+        apiVersion:
+          type: string
+        bastion:
+          properties:
+            ebsOptimized:
```
Comment (tariq1890) [inline config/crds/azureprovider_v1alpha1_azureclusterproviderstatus.yaml]: AWS specific I assume?

Comment (justaugustus) [reply config/crds/azureprovider_v1alpha1_azureclusterproviderstatus.yaml]: Yep, I'll be cleaning up the types in a separate PR!

File: cmd/versioninfo/BUILD (hunk 9 above)
Comment (justaugustus) [reply cmd/versioninfo/BUILD]: Yep! Working on adding bazel in a separate branch: https://github.com/justaugustus/cluster-api-provider-azure/tree/build

I'll clean the bazel references up in this one.

File: pkg/deployer/deployer.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+		return scope.ClusterStatus.Network.APIServerIP.IPAddress, nil
+	}
+
+	pipsvc := network.NewService(scope)
+
+	pip, err := pipsvc.GetPublicIPAddress(scope.ClusterConfig.ResourceGroup, resources.GetPublicIPName(machine))
+	if err != nil {
+		return "", err
+	}
+	return *pip.IPAddress, nil
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply pkg/deployer/deployer.go]: Yep, right now, capz (my new affectionate name for this project) only deploys a public master node, so we need to key on it's IP. Once this refactor lands, I'm going to look into adding LB, bastion, node, etc.

File: pkg/record/BUILD [hunk 13]
```diff
@@ -0,0 +1,13 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library")
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply pkg/record/BUILD]: Thanks for the tip, @vincepri!

File: cmd/versioninfo/BUILD (hunk 9 above)
Comment (justaugustus) [reply cmd/versioninfo/BUILD]: Removed!

File: cmd/clusterctl/examples/azure/README.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+For convenience, a generation script which populates templates based on azure configuration is provided.
+
+1. Run the generation script.
+
+``` shell
+./generate-yaml.sh
+```
+
+If yaml file already exists, you will see an error like the one below:
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/README.md]:
```suggestion
If the yaml file already exists, you will see an error like the one below:
```

File: pkg/deployer/deployer.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (soggiest) [reply pkg/deployer/deployer.go]: Yup, waiting on this PR to land before I added the LB services

**Comment by soggiest** [comment]
/lgtm
This will unlock a lot of work for us

File: README.md [hunk 15]
```diff
@@ -43,36 +68,66 @@ An alternative is to install [Azure CLI](https://docs.microsoft.com/en-us/cli/az
...
+TAG=0.2.0-alpha.3 \
+make docker-push
+```
+
+**NOTE:** In order for the created images to be used for testing, you must push them to a public container registry.
+
+### Submitting PRs and testing
+
+Pull requests and issues are highly encouraged!
+If you're interested in submitting PRs to the project, please be sure to run to run some initial checks prior to submission:
```
Comment (justaugustus) [inline README.md]:
```suggestion
If you're interested in submitting PRs to the project, please be sure to run some initial checks prior to submission:
```

File: cmd/clusterctl/examples/azure/README.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+For convenience, a generation script which populates templates based on azure configuration is provided.
+
+1. Run the generation script.
+
+``` shell
+./generate-yaml.sh
+```
+
+<<<<<<< HEAD
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/README.md]: Unresolved merge conflict?

File: cmd/clusterctl/examples/azure/machines.yaml.template [hunk 17]
```diff
@@ -1,60 +1,63 @@
...
\ No newline at end of file
+  - apiVersion: "cluster.k8s.io/v1alpha1"
+    kind: Machine
+    metadata:
+      name: controlplane-0
+      labels:
+        set: controlplane
+    spec:
+      versions:
+        kubelet: 1.12.5
```
Comment (tariq1890) [inline cmd/clusterctl/examples/azure/machines.yaml.template]: This should be `kustomizable` too no?

File: cmd/clusterctl/examples/azure/README.md (hunk 16 above)
Comment (justaugustus) [reply cmd/clusterctl/examples/azure/README.md]: Weird this didn't show on my local. Will fix.

File: Makefile [hunk 18]
```diff
@@ -13,9 +13,9 @@
 # limitations under the License.
 
 # Image URL to use all building/pushing image targets
-PREFIX ?= platform9
-NAME = cluster-api-azure-provider-controller
-TAG ?= latest
+PREFIX ?= quay.io/k8s
+NAME ?= cluster-api-azure-controller
+TAG ?= 0.2.0-alpha.3
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply Makefile]: Denotes a pre-release in semver: https://semver.org/#spec-item-9
Once this merges, I'll update to `0.2.0`, tag/cut our first actual release of the project, and push a new image tag. :)

File: pkg/deployer/deployer.go [hunk 19]
```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline pkg/deployer/deployer.go]:
```suggestion
Copyright 2019 The Kubernetes Authors.
```

File: pkg/cloud/azure/actuators/machine/annotations.go [hunk 20]
```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline pkg/cloud/azure/actuators/machine/annotations.go]:
```suggestion
Copyright 2019 The Kubernetes Authors.
```

File: pkg/cloud/azure/actuators/machine_scope.go [hunk 21]
```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (tariq1890) [inline pkg/cloud/azure/actuators/machine_scope.go]:
```suggestion
Copyright 2019 The Kubernetes Authors.
```

File: README.md [hunk 22]
```diff
@@ -1,23 +1,48 @@
...
-6. Build the `clusterctl` tool.
 
-   ```bash
-   make clusterctl
-   ```
+### Requirements
+
+- Linux or MacOS (Windows isn't supported at the moment)
+- A set of Azure credentials sufficient to bootstrap the cluster (an Azure service principal with Collaborator rights).
+- [KIND](https://sigs.k8s.io/kind)
```
Comment (tariq1890) [inline README.md]: In the interest of consistency, let's move the url below the list and add the url reference here instead.

File: README.md [hunk 23]
```diff
@@ -1,23 +1,48 @@
...
+- [kustomize][kustomize]
+- make
+- gettext (with `envsubst` in your PATH)
+- bazel
+
+### Optional
+
+- [Homebrew][brew] (MacOS)
+- [jq][jq]
+- [Go](https://golang.org/dl/)
```
Comment (tariq1890) [inline README.md]: Same as previous comment

File: README.md [hunk 24]
```diff
@@ -43,36 +68,66 @@ An alternative is to install [Azure CLI](https://docs.microsoft.com/en-us/cli/az
...
+   **NOTE:** Kubernetes Version >= 1.11 is required to enable CRD subresources without needing a feature gate.
 
     **Linux**
     ```bash
-    ./bin/clusterctl create cluster --provider azure \
-    -m cmd/clusterctl/examples/azure/out/machines.yaml \
-    -c cmd/clusterctl/examples/azure/out/cluster.yaml \
-    -p cmd/clusterctl/examples/azure/out/provider-components.yaml \
-    --vm-driver kvm2 --minikube kubernetes-version=v1.12.2
+    ./bin/clusterctl create cluster -v 3 \
```
Comment (tariq1890) [inline README.md]: Looks like the commands are the same for both mac and linux now. (Thanks to Kind!) . Let's just keep one :)

File: README.md [hunk 25]
```diff
@@ -43,36 +68,66 @@ An alternative is to install [Azure CLI](https://docs.microsoft.com/en-us/cli/az
...
+make docker-build
+
+# Push the image.
+PREFIX=quay.io/k8s \
+NAME=cluster-api-azure-controller \
+TAG=0.2.0-alpha.3 \
+make docker-push
+```
+
+**NOTE:** In order for the created images to be used for testing, you must push them to a public container registry.
```
Comment (tariq1890) [inline README.md]: If this is the case, we should be exposing `REGISTRY` as a user configurable variable. (This can be done in another PR)

File: cmd/clusterctl/examples/azure/machines.yaml.template (hunk 17 above)
Comment (tariq1890) [reply cmd/clusterctl/examples/azure/machines.yaml.template]: Sure :). Thanks for the explanation!

**Comment by tariq1890** [comment]
Thanks for this @justaugustus . I'm done with my review. This should be good to go after the comments are addressed.

**Comment by tariq1890** [comment]
After this PR is merged, we need to add tests back. Let's add an issue to track that and treat it as high priority.

File: README.md (hunk 25 above)
Comment (justaugustus) [reply README.md]: This can already be configured with `PREFIX` and `NAME` in the Makefile.

Comment (justaugustus) [reply README.md]: Although, maybe we could clean that up. I can assess it when I take a look at adding bazel and updating the Makefile.

File: config/manager/manager.yaml [hunk 26]
```diff
@@ -41,67 +41,70 @@ spec:
...
-        operator: Exists
-      - effect: NoExecute
-        key: node.alpha.kubernetes.io/notReady
-        operator: Exists
-      - effect: NoExecute
-        key: node.alpha.kubernetes.io/unreachable
-        operator: Exists
+        - effect: NoSchedule
+          key: node-role.kubernetes.io/master
+        - key: CriticalAddonsOnly
```
[the remark this answers is not in the document]
Comment (justaugustus) [reply config/manager/manager.yaml]: This looks right to me...

**Comment by justaugustus** [comment]
@tariq1890 -- I've addressed all the comments, this is ready for a `/lgtm` now!
/hold cancel

**Comment by justaugustus** [comment]
(Added an issue to rewrite the cluster and machine actuator tests: #68)

File: README.md (hunk 25 above)
Comment (tariq1890) [reply README.md]: Sounds good.

File: config/manager/manager.yaml (hunk 26 above)
[the remark this answers is not in the document]
Comment (tariq1890) [reply config/manager/manager.yaml]: Yes, you're right. False alarm. My bad.
